Document Set Digest functions (#20676)
prestodb/presto: presto-docs/src/main/sphinx/functions/setdigest.rst

====================
Set Digest functions
====================

MinHash, or the min-wise independent permutations locality-sensitive hashing scheme,
is a technique used in computer science to quickly estimate how similar two sets are.
MinHash serves as a probabilistic data structure that estimates the Jaccard similarity
coefficient - the measure of the overlap between two sets as a percentage of the total unique elements in both sets.
Presto offers several functions that deal with the

> `MinHash <https://wikipedia.org/wiki/MinHash>`technique.

Reviewer: ```suggestion
`MinHash <https://wikipedia.org/wiki/MinHash>`_ technique.
```

MinHash is used to quickly estimate the
`Jaccard similarity coefficient <https://wikipedia.org/wiki/Jaccard_index>`
between two sets.
It is commonly used in data mining to detect near-duplicate web pages at scale.
By using this information, the search engines efficiently avoid showing
within the search results two pages that are nearly identical.

In PrestoDB, set digest functions are a series of functions designed to create a theta sketch for specific set operations.
Set digest functions can be considered a probabilistic data structure. These functions are useful for operations
like union, intersection, and difference of sets, estimation of set sizes, and so forth.The primary set digest function
in PrestoDB is setdigest(). It combines values into a theta sketch, which is a type of
data sketch used in approximating the total distinct count of a dataset.

Here's how you use it:
SELECT cardinality(setdigest(value)) FROM table

> This generates a theta sketch and estimates the cardinality (distinct count) of the 'value' column in the 'table'.

Reviewer: ```suggestion
This generates a theta sketch and estimates the cardinality (distinct count) of the ``value`` column in the ``table``.
```

Data structures
---------------

Presto implements Set Digest data sketches by encapsulating the following components:

- `HyperLogLog <https://wikipedia.org/wiki/HyperLogLog>`
- `MinHash with a single hash function <http://wikipedia.org/wiki/MinHash#Variant_with_a_single_hash_function>`

As of now, HyperLogLog and MinHash are among the techniques implemented in Presto or used

> by certain functions in Presto to handle large data sets.

Reviewer: ```suggestion
by certain functions in Presto to handle large data sets.

```

HyperLogLog (HLL): HyperLogLog is an algorithm used to estimate the cardinality
of a set — that is, the number of distinct elements in a large data set.
Presto uses it to provide the function approx_distinct which can be used to estimate the number
of distinct entries in a column.

MinHash: MinHash is used to estimate the similarity between two or more sets, commonly known as Jaccard similarity.
It is particularly effective when dealing with large data sets and is generally used in data clustering
and near-duplicate detection.

> In the future, Presto may incorporate more sketching techniques or data structures for handling large

Reviewer: In general, documentation tends against forward-looking or future statements, and tries to focus on the existing state of the code or product. Because of this, I suggest deleting this paragraph.

        FROM (VALUES 1, 2, 3) T(value);

    Create a ``setdigest`` corresponding to a ``varchar`` array::

        SELECT make_set_digest(value)
        FROM (VALUES 'Trino', 'SQL', 'on', 'everything') T(value);
    
    Error : Unexpected parameters (varchar(10)) for function make_set_digest. 
    Expected: make_set_digest(bigint)

.. function:: merge_set_digest(setdigest) -> setdigest

Returns the ``setdigest`` of the aggregate union of the individual ``setdigest``
Set Digest structures.

.. function:: cardinality(setdigest) -> bigint

Returns the cardinality of the set digest from its internal
``HyperLogLog`` component.

    Examples::

        SELECT cardinality(make_set_digest(value))
        FROM (VALUES 1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 5) T(value);
        -- 5

.. function:: intersection_cardinality(x,y) -> bigint

Returns the estimation for the cardinality of the intersection of the two set digests.

``x`` and ``y``  be of type  ``setdigest``

    Examples::

        SELECT intersection_cardinality(make_set_digest(v1), make_set_digest(v2))
        FROM (VALUES (1, 1), (NULL, 2), (2, 3), (3, 4)) T(v1, v2);
        -- 3

.. function:: jaccard_index(x, y) -> double

> Returns the estimation of `Jaccard index <https://wikipedia.org/wiki/Jaccard_index>`for

Reviewer: ```suggestion
Returns the estimation of `Jaccard index <https://wikipedia.org/wiki/Jaccard_index>`_ for
```

the two set digests.

``x`` and ``y`` be of type  ``setdigest``.

    Examples::

        SELECT jaccard_index(make_set_digest(v1), make_set_digest(v2))
        FROM (VALUES (1, 1), (NULL,2), (2, 3), (NULL, 4)) T(v1, v2);
        -- 0.5

.. function:: hash_counts(x) -> map(bigint, smallint)

> Returns a map containing the `Murmur3Hash128 <https://wikipedia.org/wiki/MurmurHash#MurmurHash3>`

Reviewer: ```suggestion
Returns a map containing the `Murmur3Hash128 <https://wikipedia.org/wiki/MurmurHash#MurmurHash3>`_ 
```
